### PR TITLE
various otel metrics (part 1)

### PR DESCRIPTION
**Pull request by pimeys**
First part of the various metrics. Will not close the issue, and will create the second part tomorrow.

### CODE REVIEW

File: gateway/crates/gateway-binary/tests/integration_tests.rs [hunk 1]
```diff
@@ -29,6 +29,44 @@ use wiremock::{
...
+struct BatchQuery {
+    query: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    variables: Option<serde_json::Value>,
+}
+
+#[must_use]
+pub struct GqlBatchBuilder<Response> {
+    queries: Vec<BatchQuery>,
+    // These won't
```
Comment (tomhoule) [inline gateway/crates/gateway-binary/tests/integration_tests.rs:42]: Incomplete comment?

File: engine/crates/engine-v2/src/engine.rs [hunk 2]
```diff
@@ -456,6 +463,41 @@ impl<'ctx, R: Runtime> PreExecutionContext<'ctx, R> {
...
+                    document: Some(operation.prepared.metrics_attributes.sanitized_query.clone()),
+                    success: true,
+                };
+
+                self.operation_metrics.record_preparation_latency(attributes, duration);
+
+                Ok(operation)
+            }
+            Err(e) => {
+                dbg!(1);
```
Comment (tomhoule) [inline engine/crates/engine-v2/src/engine.rs:484]: stray dbg

File: engine/crates/runtime-local/src/hooks.rs [hunk 3]
```diff
@@ -20,15 +27,89 @@ struct HooksWasiInner {
...
+        let result = hook.await;
+        let duration = SystemTime::now().duration_since(start).unwrap_or_default();
+
+        let status = match result {
+            Ok(_) => HookStatus::Success,
+            Err(wasi_component_loader::Error::Internal(_)) => HookStatus::HostError,
+            Err(wasi_component_loader::Error::Guest(_)) => HookStatus::GuestError,
+        };
+
+        let attributes = vec![
```
Comment (tomhoule) [inline engine/crates/runtime-local/src/hooks.rs:48]: I think you can skip the allocation here, if you're just passing `&attributes` in the next statement.

File: engine/crates/runtime-local/src/hooks.rs [hunk 4]
```diff
@@ -20,15 +27,89 @@ struct HooksWasiInner {
...
+        let duration = SystemTime::now().duration_since(start).unwrap_or_default();
+
+        let status = match result {
+            Ok(ref statuses) if statuses.iter().any(|s| s.is_err()) => HookStatus::GuestError,
+            Ok(_) => HookStatus::Success,
+            Err(wasi_component_loader::Error::Internal(_)) => HookStatus::HostError,
+            Err(wasi_component_loader::Error::Guest(_)) => HookStatus::GuestError,
+        };
+
+        let attributes = vec![
```
Comment (tomhoule) [inline engine/crates/runtime-local/src/hooks.rs:77]: same here

File: gateway/crates/gateway-binary/tests/telemetry/metrics/operation/hooks.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
+use crate::telemetry::metrics::{with_custom_gateway, ExponentialHistogramRow, METRICS_DELAY};
+
+#[test]
+fn on_gateway_request_success() {
+    let config = indoc::indoc! {r#"
+        [hooks]
+        location = "../../../engine/crates/wasi-component-loader/examples/target/wasm32-wasip1/debug/gateway_request_no_op.wasm"
```
Comment (tomhoule) [inline gateway/crates/gateway-binary/tests/telemetry/metrics/operation/hooks.rs:7]: When are these built?

**Review by tomhoule — APPROVED** [review]

File: gateway/crates/gateway-binary/tests/telemetry/metrics/operation/hooks.rs (hunk 5 above)
Comment (pimeys) [reply gateway/crates/gateway-binary/tests/telemetry/metrics/operation/hooks.rs:7]: https://github.com/grafbase/grafbase/blob/main/.github/actions/test/action.yml#L68-L74

Comment (pimeys) [reply gateway/crates/gateway-binary/tests/telemetry/metrics/operation/hooks.rs:7]: Benjamin did a similar thing in the integration tests.

**Pull request merged**